pkg/controllers/machine_controller.go

```diff
@@ -285,6 +298,109 @@ func (r *MachineReconciler) startCheckAndEnqueueMachineState(ctx context.Context
+	isDeleting, err := r.deleteMachine(ctx, log, machine)
+	switch {
+	case isDeleting:
+		return nil
+	case err != nil:
+		return fmt.Errorf("failed to delete machine: %w", err)
+	}
+	log.V(1).Info("Deleted machine")
+
+	if err := r.deleteVolumes(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine disks: %w", err)
+	}
+	log.V(1).Info("Removed machine disks")
+
+	if err := r.deleteNetworkInterfaces(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine network interfaces: %w", err)
+	}
+	log.V(1).Info("Removed network interfaces")
+
+	if err := os.RemoveAll(r.host.MachineDir(machine.ID)); err != nil {
+		return fmt.Errorf("failed to remove machine directory: %w", err)
+	}
+	log.V(1).Info("Removed machine directory")
+
+	machine.Finalizers = utils.DeleteSliceElement(machine.Finalizers, MachineFinalizer)
+	if _, err := r.machines.Update(ctx, machine); store.IgnoreErrNotFound(err) != nil {
+		return fmt.Errorf("failed to update machine metadata: %w", err)
+	}
+	log.V(1).Info("Removed Finalizer. Deletion completed")
+
+	return nil
+}
+
+func (r *MachineReconciler) deleteMachine(ctx context.Context, log logr.Logger, machine *api.Machine) (bool, error) {
+	domain := libvirt.Domain{
+		UUID: libvirtutils.UUIDStringToBytes(machine.ID),
+	}
+
+	if machine.Spec.ShutdownAt.IsZero() {
+		machine.Spec.ShutdownAt = time.Now()
+		if _, err := r.machines.Update(ctx, machine); err != nil {
+			return false, fmt.Errorf("failed to update ShutdownAt: %w", err)
+		}
+		log.V(1).Info("Updated machine ShutdownAt", "ShutdownAt", machine.Spec.ShutdownAt)
+	}
+
+	if time.Now().Before(machine.Spec.ShutdownAt.Add(r.gcVMGracefulShutdownTimeout)) {
+		// Trigger machine shutdown until VMGracefulShutdownTimeout is over
+		return r.shutdownMachine(log, machine, domain)
+	}
+
+	return false, r.destroyDomain(log, domain)
+}
+
+func (r *MachineReconciler) destroyDomain(log logr.Logger, domain libvirt.Domain) error {
+	// DomainDestroyFlags is a blocking operation, and its synchronous nature may pose potential performance issues in the future.
+	// During test involving 26 empty disks, the function call took a maximum of 1 second to complete.
+	if err := r.libvirt.DomainDestroyFlags(domain, libvirt.DomainDestroyGraceful); err != nil {
+		if libvirt.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to initiate forceful shutdown: %w", err)
+	}
+
+	log.V(1).Info("Destroyed domain")
+	return nil
+}
+
+func (r *MachineReconciler) shutdownMachine(log logr.Logger, machine *api.Machine, domain libvirt.Domain) (bool, error) {
+	log.V(1).Info("Triggering shutdown", "ShutdownAt", machine.Spec.ShutdownAt)
+	if err := r.libvirt.DomainShutdownFlags(domain, libvirt.DomainShutdownAcpiPowerBtn); err != nil {
+		if libvirt.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to initiate shutdown: %w", err)
+	}
+
```

Comment:
Add a log here that the `Shutdown` has been triggered.

Reply:
you have log before triggering. if it is end with error you will get error message. Isn't this better way how logs stuff? you know what was current operation and you don't need additional fields for errors.

Reply:
Is it necessary in this context? Given that we are already logging `Triggering shutdown`, the absence of any subsequent error logs implies it's triggered.